### PR TITLE
Generalize the project to other fuzzers: AFL and Honggfuzz

### LINKED ISSUE #34: Decide crate directory conventions and repository organization
What should the layout of this repository be?

So far, it's been:

```
/<published crate name>/{Cargo.toml, <targets>.rs}
```

With the exception of the `crypto-hashes` directory which is just a collection of different crates. What if a crate isn't published?

### LINKED ISSUE #66: When using run-fuzzer.sh script, isolate seed/corpus directories per fuzz target

### LINKED ISSUE #73: Write script that: picks a fuzz target, runs for N minutes, repeats
This would allow one to fuzz all targets  indefinitely with a single script

### LINKED ISSUE #98: Add target for 'proc-macro2' crate
https://docs.rs/proc-macro2/0.1.10/proc_macro2/struct.TokenStream.html

### LINKED ISSUE #100: Generalize the project to other fuzzers: AFL and Honggfuzz
Hi, I think it would be really great to be able to use all the available Rust fuzzers as they all have their strengths and weaknesses.

I started a demo project inspired by this project:
https://github.com/PaulGrandperrin/fuzz-targets-rs

Any feedback is welcome!

### PR DESCRIPTION
**Pull request by PaulGrandperrin**
This is far from perfect (all targets are compiled at once) but it works and it's easy to work with (look at and add targets).

All the targets have been ported with only 1 or 2 exceptions that were broken and not easily fixable.
All seeds have been moved over too.
I added a few more targets like `url` and `proc_macro2`.

closes #100 
closes #98 
closes #66 I guess
closes #34 not relevant anymore
closes #73

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+addons:
+  apt:
+    packages:
+      - binutils-dev
+      - libunwind8-dev
+sudo: required
+notifications:
+  email:
+    recipients:
+      - paul.grandperrin@gmail.com
```
[the remark this answers is not in the document]
Comment (PaulGrandperrin) [reply .travis.yml]: maybe not :smile: , I'll see if it's too much!

File: test.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh -ve
+
+sh -ve gen-targets-src.sh
+
+cargo install afl honggfuzz --force --verbose
```
[the remark this answers is not in the document]
Comment (PaulGrandperrin) [reply test.sh]: I didn't know about `cache: cargo`. Is it really useful when always running on the latest nightly?

File: common/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+extern crate toml;
+extern crate url;
+extern crate uuid;
+extern crate xml;
+extern crate zip;
+extern crate zopfli;
+
+// many function bodies are copied from https://github.com/rust-fuzz/targets
+
+#[inline(always)]
```
[the remark this answers is not in the document]
Comment (PaulGrandperrin) [reply common/src/lib.rs]: Exactly same feeling here.
I haven't figured out what would be the best solution but I'm planning to do something about it.

I think it's ok for now.

File: .travis.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (killercup) [reply .travis.yml]: Okay, I'm pretty sure this settings means you and you alone will be responsible for our CI, though 😄

File: test.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (killercup) [reply test.sh]: Hm, good point. It'll probably only cache the crate sources. Not sure how much time the downloads tage

File: .travis.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (PaulGrandperrin) [reply .travis.yml]: I won't hesitate to ping people if needed ;-)

File: test.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (PaulGrandperrin) [reply test.sh]: Since the crates are downloaded from git, I don't think it'll change a lot but I guess there are no drawbacks in our use-case, right?

**Comment by killercup** [comment]
Just ran the continuous script on may macbook for a bit and after a while it failed to compile targets because cc failed with "posix_spawn failed: Resource temporarily unavailable". This is [apparently](https://lists.apple.com/archives/xcode-users/2012/Nov/msg00215.html) because of a hard limit of processes per user was reached. Fuzz targets don't seem to be properly killed. (I was able to solve this temporarily with `pkill -f hfuzz_target`.)

**Comment by PaulGrandperrin** [comment]
Oh, that's unexpected!
I'll try to reproduce on Linux.
Anyway, I wouldn't recommend serious fuzzing on macOS as it is less well [instrumented](https://github.com/rust-fuzz/honggfuzz-rs/blob/master/src/bin/cargo-hfuzz.rs#L160) and it seems to me that it's potentially far slower.
However I have very weak experience to back that up:
- the fuzzing part of my honggfuzz tests on travis are quite slower on macOS
- fuzzing is way slower on my high sierra guest VM than on my debian host (but that's to be expected)

If you have Linux natively installed your apple box, I would interrested in a fair benchmark ;-)

**Comment by frewsxcv** [comment]
love this, thanks so much @PaulGrandperrin! as per the previous convo, looks like there might a bug to work out, but agreed it shouldn't block this. 🎉 

bors r+

**Pull request merged**